Maven integration

fcrespo82 opened the pull request:
Opened this pull request for discussion on where to improve and fixes to do on this feature before merging into master if approved.

pflannery commented:
@fcrespo82 Thanks for the PR. 

Before I can start to looking at this please can you convert all .ts files to .js files because this isn't a typescript project. It only uses typescript to compile.

Can you also provide a smoke test in a new folder called maven under https://github.com/vscode-contrib/vscode-versionlens/tree/master/test/smoke

fcrespo82 commented:
Of course... I'll do that!

There is still work to be done before the merge... 

I'll let you know when is ready!

fcrespo82 commented:
Hey @pflannery. I think the pull request is good enough for merging.

Can you take a look and point me if there's anything else I can improve.

Thanks for your time!

pflannery commented:
@fcrespo82 After you resolve my comments above can you also merge the lastest master in to this PR and resolve the conflicts.

fcrespo82 commented:
Just merged back. All the other concerns were already resolved.
Can you check back?

fcrespo82 commented:
Just merged back. All the other concerns were already resolved.
Can you check back?

fcrespo82 commented:
Sorry, I missed that... Of course I'll do it..

pflannery commented:
@fcrespo82 It's looking good so far however I've noticed that the current selected version analysis is appearing in different places. 

![image](https://user-images.githubusercontent.com/1727302/46246319-ad7e3980-c3f3-11e8-989f-eee4eb81e56f.png)

Please make sure it always appears first in the list for each entry

![image](https://user-images.githubusercontent.com/1727302/46246307-7871e700-c3f3-11e8-8975-53849457df0f.png)

fcrespo82 commented:
Hi! I fixed that inconsistency and fixed one with tagged versions too.

Should be better now..

pflannery commented:
@fcrespo82 Everything is great. 

I noticed that you introduced a `wrapVersion` function to try and get around not having to include speech marks for child xml tags `<version>#.#.#</version>`

When this project was first setup up the json parser returned start and end offsets that included the speech marks. So when the dotnet parser came along it was implemented in the same manor but it doesn't actually have child xml `<version>` tags so we never hit the issue you found until now.

So as this project grows I think it was best that I update the json and xml parsers to exclude speech marks from the start and end offsets.

This now means that you don't need to include a `wrapVersion` in your PR anymore but it does mean there is a conflict now (shouldn't be too hard to resolve).

pflannery commented on src/common/packageCodeLens.js:
```diff
@@ -24,6 +24,12 @@ export class PackageCodeLens extends CodeLens {
     return this.package.customGenerateVersion.call(this, this.package, newVersion);
   }
 
+  wrapVersion(version) {
```
You can remove wrapVersion

pflannery commented on src/commands/factory.js:
```diff
@@ -114,7 +114,7 @@ export function createGithubCommand(codeLens) {
         `${meta.category}: ${codeLens.getInstallIndicator()} ${entry.version}`,
         `${appSettings.extensionName}.updateDependencyCommand`,
-        [codeLens, `"${newVersion}"`]
+        [codeLens, `${codeLens.wrapVersion(newVersion)}`]
```
Undo your edit wrapVersion here

pflannery commented on src/commands/factory.js:
```diff
@@ -71,7 +71,7 @@ export function createSatisfiedWithNewerCommand(serverVersion, codeLens) {
     `Satisfies ${codeLens.getInstallIndicator()} ${serverVersion}`,
     `${appSettings.extensionName}.updateDependencyCommand`,
-    [codeLens, `"${replaceWithVersion}"`]
+    [codeLens, `${codeLens.wrapVersion(replaceWithVersion)}`]
```
Undo your edit wrapVersion here

pflannery commented on src/commands/factory.js:
```diff
@@ -58,7 +58,7 @@ export function createNewVersionCommand(newVersion, codeLens) {
     `${codeLens.getTaggedVersionPrefix()}${codeLens.getInstallIndicator()} ${newVersion}`,
     `${appSettings.extensionName}.updateDependencyCommand`,
-    [codeLens, `"${replaceWithVersion}"`]
+    [codeLens, `${codeLens.wrapVersion(replaceWithVersion)}`]
```
Undo your edit wrapVersion here

pflannery commented on src/common/packageGeneration.js:
```diff
@@ -43,11 +43,12 @@ export function createInvalidVersion(name, version, type) {
 }
 
-export function createPackage(name, version, meta, customGenerateVersion) {
+export function createPackage(name, version, meta, customGenerateVersion, customWrapVersion) {
```
remove customWrapVersion here

pflannery commented on src/common/packageGeneration.js:
```diff
@@ -43,11 +43,12 @@ export function createInvalidVersion(name, version, type) {
     meta,
-    customGenerateVersion
+    customGenerateVersion,
+    customWrapVersion
```
remove customWrapVersion here

pflannery commented on src/providers/maven/mavenPackageParser.js:
```diff
@@ -0,0 +1,62 @@
+  return mavenGetPackageVersions(name)
+    .then((versions) => {
+      // console.log(versions);
+      let customWrapVersion = (v) => {
```
remove customWrapVersion here

pflannery commented on src/providers/maven/mavenPackageParser.js:
```diff
@@ -0,0 +1,62 @@
+            requestedVersion,
+            meta,
+            null,
+            customWrapVersion
```
remove customWrapVersion here

fcrespo82 commented:
Hi @pflannery I updated the branch removing the wrapVersion, added the support to use the repositories set on the current pom and also fixed the regex for tagged version and if the url didn't ends with "/". I updated the smoke test and should be good to go.

My colleagues at work are excited for this PR. They found it great for Java development.

Thanks again for the awesome extension.
